distributed/tests/test_worker_state_machine.py

```diff
@@ -285,23 +300,179 @@ async def test_fetch_to_compute(c, s, a, b):
+            await asyncio.sleep(0.01)
+
+        if as_deps:
+            y2 = c.submit(inc, x, key="y2", workers=[w1.address])
+        else:
+            s.request_acquire_replicas(w1.address, ["x"], stimulus_id="test")
+
+        while w1.tasks["x"].who_has != {w3.address}:
+            await asyncio.sleep(0.01)
+
+    await wait_for_state("x", "memory", w1)
+
+    assert_story(
+        w1.story("request-dep"),
+        [("request-dep", w3.address, {"x"})],
+        # This tests that there has been no attempt to contact w2.
+        # If the assumption being tested breaks, this will fail 50% of the times.
+        strict=True,
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_fetch_to_missing(c, s, a, b):
+    """
+    1. task x is a dependency of y
+    2. scheduler calls handle_compute("y", who_has={"x": [b]}) on a
+    3. x transitions released -> fetch -> flight; a connects to b
+    4. b responds it's busy. x transitions flight -> fetch
+    5. The busy state triggers an RPC call to Scheduler.who_has
+    6. the scheduler responds {"x": []}, because w1 in the meantime has lost the key.
+    7. x is transitioned fetch -> missing
+    """
+    x = await c.scatter({"x": 1}, workers=[b.address])
+    b.total_in_connections = 0
+    # Crucially, unlike with `c.submit(inc, x, workers=[a.address])`, the scheduler
+    # doesn't keep track of acquire-replicas requests, so it won't proactively inform a
+    # when we call remove_worker later on
+    s.request_acquire_replicas(a.address, ["x"], stimulus_id="test")
+
+    # state will flip-flop between fetch and flight every 150ms, which is the retry
+    # period for busy workers.
+    await wait_for_state("x", "fetch", a)
+    assert b.address in a.busy_workers
+
+    # Sever connection between b and s, but not between b and a.
+    # If a tries fetching from b after this, b will keep responding {status: busy}.
+    b.periodic_callbacks["heartbeat"].stop()
+    await s.remove_worker(b.address, close=False, stimulus_id="test")
+
+    await wait_for_state("x", "missing", a)
+
+    assert_story(
+        a.story("x"),
+        [
+            ("x", "ensure-task-exists", "released"),
+            ("x", "update-who-has", [], [b.address]),
+            ("x", "released", "fetch", "fetch", {}),
+            ("gather-dependencies", b.address, {"x"}),
+            ("x", "fetch", "flight", "flight", {}),
+            ("request-dep", b.address, {"x"}),
+            ("busy-gather", b.address, {"x"}),
+            ("x", "flight", "fetch", "fetch", {}),
+            ("x", "update-who-has", [b.address], []),  # Called Scheduler.who_has
+            ("x", "fetch", "missing", "missing", {}),
+        ],
+        # There may be a round of find_missing() after this.
+        # Due to timings, there also may be multiple attempts to connect from a to b.
+        strict=False,
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_self_denounce_missing_data(c, s, a):
```

Comment:
I slightly modified the current version of this test to verify it against main. I didn't like the clearing of the batched send but understand that LockedCommPool does not work since the `BatchedSend` is initialized before the pool can be installed.
I modified it such that we re-submit the cleared message and I added code to ensure that the ordering is still as expected. indeed the cluster unblocks after we resubmit the message. Therefore this test only deadlocks on main because we're dropping messages intentionally which should never happen in a real setup.

The below passes on main

```python

@gen_cluster(client=True, nthreads=[("", 1)])
async def test_self_denounce_missing_data(c, s, a):
    x = c.submit(inc, 1, key="x")
    await x

    # Wipe x from the worker. This simulates the following condition:
    # 1. The scheduler thinks a and b hold a replica of x.
    # 2. b is unresponsive, but the scheduler doesn't know yet as it didn't time out.
    # 3. The AMM decides to ask a to drop its replica.
    a.handle_remove_replicas(keys=["x"], stimulus_id="test")

    # Lose the message that would inform the scheduler.
    # In theory, this should not happen.
    # In practice, in case of TPC connection fault, BatchedSend can drop messages.
    count = a.batched_stream.batch_count
    buf = list(a.batched_stream.buffer)
    assert buf == [
        {"key": "x", "stimulus_id": "test", "op": "release-worker-data"}
    ]
    a.batched_stream.buffer.clear()

    y = c.submit(inc, x, key="y")
    # The scheduler tries computing y, but a responds that x is not available.
    # The scheduler kicks off the computation of x and then y from scratch.
    async def check_result():
        assert await y == 3

    with pytest.raises(asyncio.TimeoutError):
        await asyncio.wait_for(check_result(), 0.5)

    # There were no further messages over the batched send.
    # WE need to assert on this to make sure the ordering of events is not corrupted
    assert a.batched_stream.batch_count == count
    a.batched_stream.send(*buf)
    await check_result()
    assert_story(
        a.story("compute-task"),
        [
            ("x", "compute-task", "released"),
            # The scheduler tries computing y a first time and fails.
            # This line would not be here if we didn't lose the
            # {"op": "release-worker-data"} message earlier.
            ("y", "compute-task", "released"),
            # The scheduler receives the {"op": "missing-data"} message from the
            # worker. This makes the computation of y to fail. The scheduler reschedules
            # x and then y.
            ("x", "compute-task", "released"),
            ("y", "compute-task", "released"),
        ],
        strict=True,
    )

    del x
    while "x" in a.data:
        await asyncio.sleep(0.01)
    assert a.tasks["x"].state == "released"

```

Reply:
In the above version you can see that we're indeed blocked fetching `y` (check_result) but after the message is submitted to the scheduler, everything works.